### DISCUSSION TITLE
Output from TensorMechanicsHardeningModel

### DISCUSSION THREAD

**Discussion by Fancyshirts** [idaholab/moose#19607] [Q&A Modules: Solid mechanics] [answered]
In my application, I've written a new `TensorMechanicsHardeningModel`. It appears that at present, it is not possible to output the values returned from a `TensorMechanicsHardeningModel` directly, as they are just `GeneralUserObjects`. I would like to be able to visualise these values on the mesh, as time progresses in my transient simulation to ensure they are being computed correctly.

What is the most straightforward way of getting these properties out of the simulation? It would be nice to use the material properties output system - perhaps the best way would be to modify the `TensorMechanicsPlasticModel` I am using (Mohr-Coulomb in this case) to accept a vector of internal values and just store my `TensorMechanicsHardeningModel` parameters in the later elements of this vector? 

Unless I'm missing some already-existing functionality?

**Comment by GiudGiud** [comment]
Hello

I can see two possibly existing ways:
- the tensor mechanics action has a lot of direct output-to-exodus shortcuts. Does any of these work for you?https://mooseframework.inl.gov/syntax/Modules/TensorMechanics/Master/index.html#!
- you can implement spatialValue() for your user object, then use this AuxKernel https://mooseframework.inl.gov/moose/source/auxkernels/SpatialUserObjectAux.html to output the spatial value to an auxiliary variable. 

Guillaume

EDIT: If these quantities you are trying to access are material properties (like stress and strain), there are also auxkernels to do that. Please let us know

**Comment by hugary1995** [comment, marked as answer]
This sounds like a good place to use the `compute = false` materials (I think they are formally called "discrete materials", although I sometimes find that name a little bit confusing). Take a look at the models in [this directory](https://github.com/idaholab/moose/tree/next/modules/tensor_mechanics/src/materials/crystal_plasticity). These `compute = false` materials are very similar to userobject-based models, with a bunch of advantages.

**Reply by hugary1995** [reply]
... including outputting capabilities, e.g.
```python
[a_fancy_hardening_model]
  type = YourDiscreteMaterial
  output_properties = 'plastic_strain'
  outputs = 'exodus'
[]
```
